File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/com/example/factory/factories/DocumentFactoryRegistry.java
package com.example.factory.factories;
import java.util.HashMap;
import java.util.Map;
public class DocumentFactoryRegistry {
    private static final Map<String, DocumentFactory> factories = new HashMap<>();
    static {
        factories.put("WORD", new WordDocumentFactory());
        factories.put("PDF", new PdfDocumentFactory());
        factories.put("EXCEL", new ExcelDocumentFactory());
    }
    public static DocumentFactory getFactory(String documentType) {
        DocumentFactory factory = factories.get(documentType.toUpperCase());
        if (factory == null) {
            throw new IllegalArgumentException("Unknown document type: " + documentType);
        }
        return factory;
    }
    public static String[] getSupportedTypes() {
        return factories.keySet().toArray(new String[0]);
    }
}

File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/Factories/DocumentFactoryRegistry.cs
using System;
using System.Collections.Generic;

namespace FactoryDemo.Factories;

public static class DocumentFactoryRegistry
{
    private static readonly Dictionary<string, DocumentFactory> _factories = new(StringComparer.OrdinalIgnoreCase)
    {
        ["WORD"]  = new WordDocumentFactory(),
        ["PDF"]   = new PdfDocumentFactory(),
        ["EXCEL"] = new ExcelDocumentFactory()
    };

    public static DocumentFactory GetFactory(string documentType) =>
        _factories.TryGetValue(documentType, out var factory)
            ? factory
            : throw new ArgumentException($"Unknown document type: {documentType}", nameof(documentType));

    public static string[] GetSupportedTypes() => _factories.Keys.ToArray();
}
